### PR TITLE
MINOR: remove unnecessary timeout for admin request

### PR DESCRIPTION
**Pull request by ableegoldman**
Turns out `future.get()` actually does apply the admin's `default.api.timeout.ms` config internally, so we don't need to worry about providing a timeout of our own. Who knew

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 1]
```diff
@@ -44,12 +45,6 @@
     private final static String INTERRUPTED_ERROR_MESSAGE = "Thread got interrupted. This indicates a bug. " +
         "Please report at https://issues.apache.org/jira/projects/KAFKA or dev-mailing list (https://kafka.apache.org/contact).";
 
-    private static final class InternalAdminClientConfig extends AdminClientConfig {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:47]: Moved to `ClientUtil`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java [hunk 2]
```diff
@@ -90,24 +96,21 @@ public static String getTaskProducerClientId(final String threadClientId, final
         return result;
     }
 
-    public static Map<TopicPartition, ListOffsetsResultInfo> fetchEndOffsetsWithoutTimeout(final Collection<TopicPartition> partitions,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java:93]: Removed this as we presumably always want to apply a timeout and do not want `KafkaStreams#allLocalStorePartitionLags` to block indefinitely

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -551,6 +551,7 @@ void runLoop() {
                 if (nextProbingRebalanceMs.get() < time.milliseconds()) {
                     log.info("Triggering the followup rebalance scheduled for {} ms.", nextProbingRebalanceMs.get());
                     mainConsumer.enforceRebalance();
+                    nextProbingRebalanceMs.set(Long.MAX_VALUE);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:554]: This is neither relevant to this PR nor required for correctness, but I noticed the log message above tends to spam the logs in some tests. Since this gets set/reset at the end of every rebalance, we may as well reset it here to avoid an avalanche of `Triggering the followup rebalance...`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java [hunk 4]
```diff
@@ -90,24 +96,21 @@ public static String getTaskProducerClientId(final String threadClientId, final
...
-        return fetchEndOffsets(partitions, adminClient, null);
+    public static int getAdminDefaultApiTimeoutMs(final StreamsConfig streamsConfig) {
+        final InternalAdminClientConfig dummyAdmin = new InternalAdminClientConfig(streamsConfig.getAdminConfigs("dummy"));
+        return dummyAdmin.getInt(AdminClientConfig.DEFAULT_API_TIMEOUT_MS_CONFIG);
     }
 
     public static Map<TopicPartition, ListOffsetsResultInfo> fetchEndOffsets(final Collection<TopicPartition> partitions,
                                                                              final Admin adminClient,
-                                                                             final Duration timeout) {
+                                                                             final long timeoutMs) {
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: It seems to me ```Duration``` is more readable than ```long```. Is there a reason to make this change?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 5]
```diff
@@ -1246,7 +1247,12 @@ public void cleanUp() {
         }
 
         log.debug("Current changelog positions: {}", allChangelogPositions);
-        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets = fetchEndOffsetsWithoutTimeout(allPartitions, adminClient);
+        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets =
+            fetchEndOffsets(
+                allPartitions,
+                adminClient,
+                getAdminDefaultApiTimeoutMs(config)
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: The admin configs is built in ```KafkaStreams``` construction. Could we reuse it?

```scala
        adminClient = clientSupplier.getAdmin(config.getAdminConfigs(ClientUtils.getSharedAdminClientId(clientId)));
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java [hunk 6]
```diff
@@ -90,24 +96,21 @@ public static String getTaskProducerClientId(final String threadClientId, final
         return result;
     }
 
-    public static Map<TopicPartition, ListOffsetsResultInfo> fetchEndOffsetsWithoutTimeout(final Collection<TopicPartition> partitions,
-                                                                                           final Admin adminClient) {
-        return fetchEndOffsets(partitions, adminClient, null);
+    public static int getAdminDefaultApiTimeoutMs(final StreamsConfig streamsConfig) {
+        final InternalAdminClientConfig dummyAdmin = new InternalAdminClientConfig(streamsConfig.getAdminConfigs("dummy"));
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: Could you add comment for the "dummy"?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java [hunk 7]
```diff
@@ -40,9 +41,14 @@
 import org.slf4j.LoggerFactory;
 
 public class ClientUtils {
-
     private static final Logger LOG = LoggerFactory.getLogger(ClientUtils.class);
 
+    static final class InternalAdminClientConfig extends AdminClientConfig {
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: How about QuietAdminClientConfig, like QuietStreamsConfig?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java (hunk 4 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: I think it's because we're now also calling it right after calling `getAdminDefaultApiTimeoutMs`, so it seems a bummer to create a Duration from millis and then immediately convert it back to millis.

File: streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java [hunk 8]
```diff
@@ -89,7 +89,7 @@ public void shouldProperlyConfigureTheAssignor() throws NoSuchFieldException, Il
                 mkEntry(StreamsConfig.MAX_WARMUP_REPLICAS_CONFIG, "7"),
                 mkEntry(StreamsConfig.PROBING_REBALANCE_INTERVAL_MS_CONFIG, "480000"),
                 mkEntry(StreamsConfig.InternalConfig.ASSIGNMENT_LISTENER, configuredAssignmentListener),
-                mkEntry(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG, 9)
+                mkEntry(AdminClientConfig.DEFAULT_API_TIMEOUT_MS_CONFIG, 90_000)
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java]: Just curious, why the change from 9 to 90,000?

**Review by vvcephei — COMMENTED** [review]
LGTM! Just a few minor comments.

File: streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java]: For fun

Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java]: I mean, because the value has to be larger than the `request.timeout.ms` which defaults to 30,000

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: Good call

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: Yep

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 5 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I did it this way since we need to get the admin's default.api.timeout in other places where we only have the `streamsConfig`, so we may as well just pass that as the argument to `getAdminDefaultApiTimeoutMs`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ClientUtils.java]: Ack

**Comment by vvcephei** [comment]
Test this please

File: streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java (hunk 8 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/TaskAssignorIntegrationTest.java]: Ah, thanks.

**Review by vvcephei — APPROVED** [review]

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:554]: enforceRebalance is guaranteed not to actually run the assignment logic, right? That will only run during a call to poll, I'm hoping. Otherwise, this line should go before the call.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:554]: Yep. It just provides a notice to the consumer to enforce that a rebalance will occur on the next poll

**Comment by vvcephei** [comment]
It looks like all the review comments are addressed, and all the tests passed, so I'll proceed to merge.

**Comment by vvcephei** [comment]
Oh, actually @ableegoldman , it looks like there was a conflict with the other PR I just merged.

**Comment by vvcephei** [comment]
I just resolved the conflicts.

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Ok to test

**Comment by vvcephei** [comment]
Retest this please

**Comment by vvcephei** [comment]
... or not

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Retest this please

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
\o/

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ClientUtilsTest.java [hunk 9]
```diff
@@ -45,42 +50,42 @@ public void fetchEndOffsetsShouldRethrowRuntimeExceptionAsStreamsException() {
         final Admin adminClient = EasyMock.createMock(AdminClient.class);
         EasyMock.expect(adminClient.listOffsets(EasyMock.anyObject())).andThrow(new RuntimeException());
         replay(adminClient);
-        assertThrows(StreamsException.class, () ->  fetchEndOffsetsWithoutTimeout(emptyList(), adminClient));
+        assertThrows(StreamsException.class, () ->  fetchEndOffsets(emptyList(), adminClient, 60_000L));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ClientUtilsTest.java]: Should we pass in `MAX_VALUE` to avoid introducing test flakyness?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ClientUtilsTest.java [hunk 10]
```diff
@@ -45,42 +50,42 @@ public void fetchEndOffsetsShouldRethrowRuntimeExceptionAsStreamsException() {
...
-    public void fetchEndOffsetsShouldRethrowInterruptedExceptionAsStreamsException() throws InterruptedException, ExecutionException {
+    public void fetchEndOffsetsShouldRethrowInterruptedExceptionAsStreamsException() throws Exception {
         final Admin adminClient = EasyMock.createMock(AdminClient.class);
         final ListOffsetsResult result = EasyMock.createNiceMock(ListOffsetsResult.class);
         final KafkaFuture<Map<TopicPartition, ListOffsetsResultInfo>> allFuture = EasyMock.createMock(KafkaFuture.class);
 
         EasyMock.expect(adminClient.listOffsets(EasyMock.anyObject())).andStubReturn(result);
         EasyMock.expect(result.all()).andStubReturn(allFuture);
-        EasyMock.expect(allFuture.get()).andThrow(new InterruptedException());
+        EasyMock.expect(allFuture.get(60000L, TimeUnit.MILLISECONDS)).andThrow(new InterruptedException());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ClientUtilsTest.java]: As above (also below)

Also nit: `60_000L` (if just `60` and `TimeUnit.SECONDS`?)

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 5 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: We should update the JavaDocs that this method may throw a `TimeoutException` now. What make we wondering if this is a public API change? Was there any discussion on the original KIP about the behavior of `allLocalStorePartitionLags` ?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: This change might require a KIP... \cc @vvcephei @guozhangwang WDYT?

**Review by vvcephei — APPROVED** [review]
Thanks for the update @ableegoldman

**Comment by vvcephei** [comment]
Test this please

**Comment by vvcephei** [comment]
Test this please

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 11]
```diff
@@ -1246,7 +1246,7 @@ public void cleanUp() {
         }
 
         log.debug("Current changelog positions: {}", allChangelogPositions);
-        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets = fetchEndOffsetsWithoutTimeout(allPartitions, adminClient);
+        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets = fetchEndOffsets(allPartitions, adminClient);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1250]: Can we please update the JavaDocs of `allLocalStorePartitionLags` to state that a `StreamsException` could be thrown?

**Review by mjsax — APPROVED** [review]

**Comment by mjsax** [comment]
Checkstyle error:
```
[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.12/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:21:8: Unused import - org.apache.kafka.clients.admin.AdminClientConfig. [UnusedImports]
```

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 11 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1250]: ack, done

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.6`. Do we want this in `2.5`, too? If yes, we would need a new PR. Not possible to cherry-pick.

**Comment by ableegoldman** [comment]
Thanks! I actually don't think there ended up being anything relevant t o2.5 in the final form of this PR. Except maybe adding `@throws StreamsException` to the `allLocalStorePartitionLags` javadocs, but not sure that warrants an entire PR

**Comment by mjsax** [comment]
Ack. In 2.5 we use the AdminClient directly in `allLocalStorePartitionLags` and don't apply a timeout on `get()`. -- Might still be worth to do quick PR to update the JavaDocs :)

**Comment by ableegoldman** [comment]
Fair enough: https://github.com/apache/kafka/pull/8772
